feat: add cache.storeFilter for filtering cached assets

cseas opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

Fixes https://github.com/webpack/webpack/issues/19359

<!-- In addition to that please answer these questions: -->

**What kind of change does this PR introduce?**

<!-- E.g. a bugfix, feature, refactoring, build related change, etc… -->

Add a new config option `cache.storeFiles` which gives the users an option to configure filtering for which assets should be excluded from cache.

**Did you add tests for your changes?**

<!-- Note that we won't merge your changes if you don't add tests -->

Yes

**Does this PR introduce a breaking change?**

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

No

**What needs to be documented once your changes are merged?**

- The new option needs to be documented here: https://webpack.js.org/configuration/cache/
- Might need to change the part about `cache.type: 'memory'` not allowing additional configuration.

<!-- List all the information that needs to be added to the documentation after merge -->
<!-- When your changes are merged you will be asked to contribute this to the documentation -->

cseas committed:
feat: add cache.storeFilter for filtering cached assets

cseas commented on declarations/WebpackOptions.d.ts:
```diff
@@ -1040,7 +1040,7 @@ export interface WebpackOptions {
 export interface MemoryCacheOptions {
 	/**
-	 * Additionally cache computation of modules that are unchanged and reference only unchanged modules.
+	 * Additionally cache computation of modules that are unchanged and reference only unchanged modules in memory.
```
I didn't change this line. This was added automatically by `yarn run fix:special`

cseas commented on declarations/WebpackOptions.d.ts:
```diff
@@ -1133,6 +1133,10 @@ export interface FileCacheOptions {
+	/**
+	 * A function to filter which cache entries should be stored. Return false to skip storing a cache entry.
+	 */
+	storeFilter?: (identifier: string, data: any) => boolean;
```
Function signature is similar to the one suggested by @alexander-akait [here](https://github.com/webpack/webpack/issues/19359#issuecomment-2755764498) but I'm not sure if there are any use-cases for filtering on `data`. 

Need input from the maintainers here. If it isn't required, I'll go ahead and remove the `data` parameter.

cseas committed:
refactor: typo

alexander-akait reviewed (commented):
It should work for memory cache too, also please add test cases

cseas committed:
feat: add storeFilter to get method of IdleFileCachePlugin

cseas committed:
feat: add storeFilter to MemoryCachePlugin

cseas committed:
test: MemoryCachePlugin storeFilter

cseas committed:
fix: typedefs

cseas committed:
test: IdleFileCachePlugin

cseas commented:
> It should work for memory cache too, also please add test cases

@alexander-akait I've added the storeFilter property to MemoryCachePlugin as well. Also added unit tests for both IdleFileCachePlugin and MemoryCachePlugin.

alexander-akait commented:
Close in favor of https://github.com/webpack/webpack/pull/21237
